### PR TITLE
fix(moe): write fused-shared routing replay as routed-only [T, K]

### LINKED ISSUE #5009: [Bug] routing_replay_out out-of-bounds writes: fused-shared stride and missing dim0 lower bound
Two out-of-bounds writes on `routing_replay_out` in the flat `trtllm_*_moe` APIs. Both only touch that optional buffer (`if (mPtrRoutingReplayOut != nullptr)`). Default `None` / unified `MoELayer` and MoE numerics are unaffected.

`routing_replay_out` is `[num_tokens, top_k]` `int16`: the routed top-k expert ids from FromLogits routing. Fused shared slots are not stored there.

## 1. Fused shared experts: replay indexed as if it were packed `[T, K+S]`

**Where:** `routingMainKernel` in `csrc/fused_moe/trtllm_backend/trtllm_fused_moe_routing_deepseek.cu`.

DeepSeek packed ids/weights are `[T, K+S]`: each token has `K` routed slots then `S` fused-shared slots (`mTotalExpertsPerToken = top_k + num_fused_shared_experts`). Replay is a different tensor, `[T, K]` — routed ids only.

The replay store reused the packed index:

```cpp
auto idxTopK = blockIdx.x * params.mTotalExpertsPerToken + laneIdx;  // token * (K+S) + k
params.mPtrRoutingReplayOut[idxTopK] = static_cast<int16_t>(expertIdx);
```

That is correct for packed ids. For replay it is `S` extra elements per token. Token 0 lands in the right place; token 1 starts at element `K+S` instead of `K`, overlapping token 0's row; later tokens walk off the end of the `[T, K]` allocation.

`S == 0` → `K+S == K`, so the bug is invisible. Only DeepSeek sets `mNumFusedSharedExperts`; custom/llama4 already index replay with `* mTopK`.

Host and C++ rejected `routing_replay_out` with `S > 0` (#4239), so this OOB was unreachable and the combination was unavailable.

**Fix:** index replay with `blockIdx.x * mTopK + laneIdx`. Leave packed/weights on `K+S`. Drop the `S > 0` rejects. Replay stays `[T, K]`; passing `[T, K+S]` should still fail `dim1 == top_k`.

**PR:** #4894

## 2. Undersized dim0 accepted, written past the end

**Where:** `validate_routing_replay_out` / `_validate_routing_replay_out`.

The kernel writes one replay row per token (DeepSeek: `row = blockIdx.x`). Neither validator required `dim0 >= num_tokens`. The C++ comment said dim0 was unchecked for CUDA graphs; graphs need *oversized* buffers to stay legal, not the lower bound dropped.

**Repro:** `num_tokens=1024` against `[8, 2]`. `compute-sanitizer --tool memcheck` with `PYTORCH_NO_CUDA_MEMORY_CACHING=1`:

```
Invalid __global__ write of size 2 bytes
    at routingDeepSeek::routingMainKernel<...>+0x2850
    by thread (0,0,0) in block (16,0,0)
    Access ... is out of bounds
    and is 33 bytes after the nearest allocation ... of size 32 bytes
```

With the caching allocator this is silent corruption of a neighbouring tensor. vLLM's max-batch pre-allocation does not hit it.

**Fix:** `dim0 >= num_tokens` in both validators; oversized still legal.

**PR:** #5072 (merged)

### PR DESCRIPTION
**Pull request by feih-nv**
## 📌 Description

Fixes #5009 part 1. DeepSeek `routingMainKernel` wrote `routing_replay_out` with the packed index `token * (K+S) + k`. Replay is `[T, K]` (routed ids only), so token 1 starts at `K+S` instead of `K`. `S == 0` hid it; host/C++ rejected `S > 0` (#4239), so the combo was unavailable.

This PR indexes replay with `token * K + k` and drops the `S > 0` rejects. Packed ids/weights stay `[T, K+S]`.

## 🔍 Related Issues

- https://github.com/flashinfer-ai/flashinfer/issues/5009
- https://github.com/flashinfer-ai/flashinfer/pull/5072
- https://github.com/flashinfer-ai/flashinfer/pull/4239

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

### CPU reject
- `dim1 == K+S`
- non-DeepSeek + `S > 0`

### GPU (SM100) accept
- FP8 / FP4 fused-shared replay, `T ∈ {8,32}`, `S ∈ {1,2}`
- one FP8 CUDA-graph capture on oversized `[32, K]`

## Reviewer Notes

With `S > 0`, `routing_replay_out` is still `[T, K]` `int16`: the same `K` routed expert ids as `S = 0` on the same logits, ids in `[0, E)`. The `S` fused-shared slots live only in the internal packed ids/weights (`[T, K+S]`, ids `E … E+S-1`). They are not appended to replay. A `[T, K+S]` replay buffer is rejected (`dim1` must equal `top_k`).

## Summary by CodeRabbit

* **New Features**
  * Routing replay is now supported with fused shared experts for FP8 and FP4 MoE operations.
  * Replay buffers retain a routed-only layout with width equal to `top_k`; shared-expert slots are not recorded.
  * Support includes oversized buffers and CUDA graph execution.

* **Documentation**
  * Updated integration and testing guidance for routed-only replay behavior.

* **Tests**
  * Added coverage for replay contents, layouts, sentinel rows, and FP8/FP4 execution paths.

### CODE REVIEW

**Comment by feih-nv** [comment]
@flashinfer-bot run

**Comment by feih-nv** [comment]
@flashinfer-bot run

**Comment by feih-nv** [comment]
@flashinfer-bot run

**Comment by feih-nv** [comment]
@flashinfer-bot run

**Review by aleozlx — APPROVED** [review]
